src/bsp/src/stm32/src/SpiEsp.cpp

```diff
@@ -0,0 +1,211 @@
+
+    switch (m_rxState) {
+    case receiveState::IDLE:
+        if (m_inboundRequest || m_txState != transmitState::IDLE) {
+            rxLengthBytes = EspHeader::sizeBytes;
+            m_rxState = receiveState::RECEIVING_HEADER;
+        }
+        break;
+    case receiveState::RECEIVING_HEADER:
+        rxLengthBytes = EspHeader::sizeBytes;
+        break;
+    case receiveState::PARSING_HEADER:
+        m_inboundHeader = (EspHeader::Header*)m_inboundMessage.m_data.data();
+        if (m_inboundHeader->crc8 !=
+            m_crc.calculateCRC8(m_inboundHeader, EspHeader::sizeBytes - 1)) {
+            m_logger.log(LogLevel::Error, "Received corrupted header");
+            m_logger.log(LogLevel::Debug, "Bytes were: | %d | %d | %d | %d |",
+                         m_inboundMessage.m_data[0], m_inboundMessage.m_data[1],
+                         m_inboundMessage.m_data[2], m_inboundMessage.m_data[3]);
+            m_rxState = receiveState::ERROR;
+            break;
+        }
+        if (m_inboundHeader->rxSizeWord << 2 == m_outboundMessage.m_sizeBytes &&
+            m_outboundMessage.m_sizeBytes != 0) {
+            m_logger.log(LogLevel::Debug, "Received valid header. Can now send payload");
+            m_txState = transmitState::SENDING_PAYLOAD;
+        }
+        else if (m_inboundHeader->rxSizeWord == 0 && !m_inboundHeader->systemState.espSystemState.failedCrc && m_outboundMessage.m_sent) {
+            m_outboundMessage.m_sizeBytes = 0;
+            m_txState = transmitState::SENDING_HEADER;
+        }
+        else {
+            m_txState = transmitState::SENDING_HEADER;
+            m_logger.log(LogLevel::Debug, "Received valid header but cannot send payload");
+        }
+        // This will be sent on next header. Payload has priority over headers.
+        m_inboundMessage.m_sizeBytes = m_inboundHeader->txSizeWord << 2;
+        if (m_inboundMessage.m_sizeBytes == m_outboundHeader.rxSizeWord << 2 &&
+            m_inboundMessage.m_sizeBytes != 0) {
+            rxLengthBytes = m_inboundHeader->txSizeWord << 2;
+            m_outboundHeader.systemState.stmSystemState.failedCrc = 0;
+            m_rxState = receiveState::RECEIVING_PAYLOAD;
+        } else {
+            rxLengthBytes = EspHeader::sizeBytes;
+            m_rxState = receiveState::RECEIVING_HEADER;
+        }
+        break;
+    case receiveState::RECEIVING_PAYLOAD:
+        m_outboundHeader.systemState.stmSystemState.failedCrc = 0;
+        rxLengthBytes = m_inboundHeader->txSizeWord << 2;
+        break;
+    case receiveState::VALIDATE_CRC:
+        if (m_crc.calculateCRC32(m_inboundMessage.m_data.data(),
+                                 m_inboundMessage.m_sizeBytes - CRC32_SIZE) !=
+            *(uint32_t*)&m_inboundMessage.m_data[m_inboundMessage.m_sizeBytes - CRC32_SIZE]) {
+            m_outboundHeader.systemState.stmSystemState.failedCrc = 1;
+        } else {
+            m_logger.log(LogLevel::Info, "ESP says: %s", m_inboundMessage.m_data.data());
```

Comment:
idk if this log is necessary since it's pretty much done in the main, it can cause issues if your data is not null terminated too